File: source/src/main/java/org/rss/tools/mpl/global/ApplicationLogger.java
package org.rss.tools.mpl.global;

import java.util.Arrays;

import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.After;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Before;
import org.aspectj.lang.annotation.Pointcut;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Logging Aspect. 
 * @author ricardo saturnino
 */
@Aspect
public class ApplicationLogger {

	private static final Logger LOG = LoggerFactory.getLogger(ApplicationLogger.class);
	
	@Pointcut("execution(* org.rss.tools.mpl.parsing..*(..))")
	public void allMethods() {}
	
	@Around("allMethods()")
	public Object aroundAll(ProceedingJoinPoint pjp) throws Throwable {
		LOG.debug("Start {}.{}", pjp.getSignature().getDeclaringType(), pjp.getSignature().getName());
		if (LOG.isTraceEnabled()) {
			LOG.trace(Arrays.toString(pjp.getArgs()));
		}

		try {
			return pjp.proceed();
		} finally {
			LOG.trace("Method {} complete", pjp.getSignature().getName());
		}
	}
	
	// ---------------------- Main Logger ------------------------

	@Before("execution(* org.rss.tools.mpl.Main.main(..))")
	public void startMain() {
		LOG.info("Starting Mockuplan process......................");
	}

	@After("execution(* org.rss.tools.mpl.Main.main(..))")
	public void allDone() {
		LOG.info("Execution ended ......................");
	}

	@Before("call(* org.rss.tools.mpl.Main.processParse(..))")
	public void startParse() {
		LOG.info("* Executing PARSE phase......................");
	}

	@Before("call(* org.rss.tools.mpl.Main.processRender(..))")
	public void startRender() {
		LOG.info("* Executing RENDER phase......................");
	}
}

File: source/src/main/kotlin/org/rss/tools/mpl/global/ApplicationLogger.kt
package org.rss.tools.mpl.global

import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.*
import org.slf4j.LoggerFactory
import java.util.*

/**
 * Logging Aspect.
 * @author ricardo saturnino
 */
@Aspect
class ApplicationLogger {

    @Pointcut("execution(* org.rss.tools.mpl.parsing..*(..))")
    fun allMethods() {
    }

    @Around("allMethods()")
    @Throws(Throwable::class)
    fun aroundAll(pjp: ProceedingJoinPoint): Any {
        LOG.debug("Start {}.{}", pjp.signature.declaringType, pjp.signature.name)
        if (LOG.isTraceEnabled) {
            LOG.trace(Arrays.toString(pjp.args))
        }

        try {
            return pjp.proceed()
        } finally {
            LOG.trace("Method {} complete", pjp.signature.name)
        }
    }

    // ---------------------- Main Logger ------------------------

    @Before("execution(* org.rss.tools.mpl.Main.main(..))")
    fun startMain() {
        LOG.info("Starting Mockuplan process......................")
    }

    @After("execution(* org.rss.tools.mpl.Main.main(..))")
    fun allDone() {
        LOG.info("Execution ended ......................")
    }

    @Before("call(* org.rss.tools.mpl.Main.processParse(..))")
    fun startParse() {
        LOG.info("* Executing PARSE phase......................")
    }

    @Before("call(* org.rss.tools.mpl.Main.processRender(..))")
    fun startRender() {
        LOG.info("* Executing RENDER phase......................")
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(ApplicationLogger::class.java)
    }
}
